Explore Metrics: Introduce augurs sorting options in breakdown view

itsmylife opened the pull request:
**What is this feature?**

**Disclaimer**
This requires the changes in [this PR](https://github.com/grafana/grafana/pull/90944)

In this PR we introduced various sorting types in label breakdown. It integrates https://github.com/grafana/augurs for sorting. 
All breakdown related files was moved to `Breakdown` directory. 

This PR was highly influenced by https://github.com/grafana/explore-logs/pull/512 kudos to @matyax 

**Why do we need this feature?**

For a comprehensive sorting solution

**Who is this feature for?**

Explore Metrics App users

**Which issue(s) does this PR fix?**:

Fixes  https://github.com/grafana/grafana/issues/90835

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

itsmylife committed:
refactor breakdown scene

itsmylife committed:
refactor BreakdownScene along with LayoutSwitcher

itsmylife committed:
rename

itsmylife committed:
don't pass default layout

itsmylife committed:
better type handling

itsmylife committed:
betterer

itsmylife committed:
add @bsull/augurs

itsmylife committed:
implement LabelBreakdownScene

itsmylife committed:
integrate SortByScene in LabelBreakdownScene

itsmylife committed:
move to new directory

itsmylife committed:
introduce BreakdownSearchScene

itsmylife committed:
integrate searchScene

itsmylife committed:
cleaning

itsmylife committed:
initialize @bsull/augurs

itsmylife committed:
add interaction

itsmylife committed:
use new breakdown scene

sd2k commented on package.json:
```diff
@@ -239,6 +239,7 @@
     "yargs": "^17.5.1"
   },
   "dependencies": {
+    "@bsull/augurs": "0.2.0",
```
Not urgent but I just released v0.3.1 of this package which has some significant perf improvements for changepoint detection, probably worth updating. It should be compatible for the APIs you're using 👍

catherineymgui commented:
Thanks for implementing this!! I added it to [figma](https://www.figma.com/design/eL7u2QrvLykWRLeUELiNNy/%5BGA%5D-Explore-Metrics---Master-File?node-id=840-15354&t=pJDcFR9HhIJYCkQp-1) and made some improvements which are also addressed below. (Also @matyax)

Feedback:

- I noticed the label on this dropdown has disappeared. Can we add it back? If anything - we can change the Sort by to match this label / field design (label on top, field below)
<img width="149" alt="Screenshot 2024-07-30 at 17 45 36" src="https://github.com/user-attachments/assets/114e5166-ec97-4646-ae04-b48a9a063588">

- In the breakdown --> all view, the grid/row is now beside the label (before it was right aligned)
<img width="427" alt="Screenshot 2024-07-30 at 17 46 22" src="https://github.com/user-attachments/assets/8ba05986-2030-4ddc-8f47-433f6dc6a8bc">

- I reworded the info tooltip a bit to tighten the language
<img width="658" alt="Screenshot 2024-07-30 at 17 41 36" src="https://github.com/user-attachments/assets/2f38e9de-3c80-4115-b3a9-6245b487143e">

- Is there a way to combine the two dropdowns into one? Similar to the dashboards sorting option. Also do we want ascending options for most of these (ex: why would someone need highest spike by ascending -- which will be lowest spike)? I designed what this could look like if we combined them. 

<img width="558" alt="Screenshot 2024-07-30 at 17 39 21" src="https://github.com/user-attachments/assets/8ded03ff-e4fc-4895-932d-1ff5aa194fd0">

- Could we give more context / description into the % options? I am not sure if I fully understand them (is it like the top % of the average values, highest value, etc.)?

- I gave this feedback to @/matryer but how do we feel about having all of these options? I know he mentioned that the “most relevant” might not be most relevant, and thus we want to provide additional flexibility, but I worry the other options may add too much confusion. Thoughts @zhehao-grafana ?

zhehao-grafana commented:
> I gave this feedback to @/matryer but how do we feel about having all of these options? I know he mentioned that the “most relevant” might not be most relevant, and thus we want to provide additional flexibility, but I worry the other options may add too much confusion.

I would only keep a few options for users like around 3, and evaluate whether we need to add more

zhehao-grafana commented:
if users select Most relevant, sorting it by Desc or Asc might not make sense, since we rank these metrics and give them the most relevant one. It makes more sense if they select Name in the dropdown
![image](https://github.com/user-attachments/assets/ec19b16d-c88b-430e-95eb-2bb9c5164ca2)

itsmylife committed:
Merge branch 'refs/heads/main' into ismail/em/implement-changepoint-sorting

# Conflicts:
#	public/app/features/trails/ActionTabs/LayoutSwitcher.tsx
#	public/app/features/trails/ActionTabs/types.ts
#	public/app/features/trails/interactions.ts
#	yarn.lock

itsmylife committed:
resolve merge conflicts

itsmylife committed:
ugrade @bsull/augurs

itsmylife committed:
update import

itsmylife committed:
update css

itsmylife committed:
update tooltip text

catherineymgui commented:
Thanks @matyax!

I'm trying to map the labels in your dashboard to the options in the dropdown. Could you provide some guidance?
StdDev --> Is this the default (most relevant)?
Changepoint --> ?
Max --> Highest spike?

matyax commented:
Sure.
- Most relevant: ML, changepoint
- Detected outliers: ML, outliers
- Widest spread: standard deviation

catherineymgui commented:
Thanks @matyax, so given the low usage on most relevant (default), does your instrumentation mean:

- Users are switching away right away to another option in the dropdown
- This is tracking clicks so if users do not click another option and remain on most relevant, this is not being tracked?

Also I'm curious about the difference between outlier sorting and the "most relevant" sorting?

Either way I think these ones make the most sense (but unsure about order):
- Most relevant (but maybe this is not first)
- Outliers
- Widest spread
- (Potentially) Name [A-Z]
- (Potentially) Name [Z-A]

But would also like your thoughts @zhehao-grafana

zhehao-grafana commented:
I might keep these four
Outliers
Widest spread
(Potentially) Name [A-Z]
(Potentially) Name [Z-A]

**Most relevant** is confusing. From user interviews, one of the main reasons that users do not use related metrics section is that they don't understand how and why they are related or relevant. I want to keep the current options as explicit as possible so users know what they are clicking into

matyax commented:
@catherineymgui I would recommend to use the instrumentation results as just another data point, because it's really challenging to understand the reasons behind those numbers. What I screenshoted tracks the default values used by users when they open the breakdowns, so, for example, a single user refreshing the page would significantly increase that metric.

These other panels (last 30 days) show what are the most popular options that the user switch to:

For labels:

![imagen](https://github.com/user-attachments/assets/7516b5d9-d22e-49a8-8458-f759ac7639cf)


For fields:

![imagen](https://github.com/user-attachments/assets/d2fc5793-a3a2-437a-bf71-e8660e787663)

itsmylife committed:
refine sorting

itsmylife committed:
fix unit test

itsmylife committed:
fix

itsmylife committed:
implement outlier detector

itsmylife committed:
support wasm

itsmylife committed:
jest testing fix

itsmylife committed:
localization fix

itsmylife committed:
use unknown instead of any

itsmylife committed:
update i18n

itsmylife committed:
update betterer

itsmylife committed:
fix locales

itsmylife committed:
update test

itsmylife committed:
fix tests maybe

itsmylife committed:
prettier

matyax committed:
chore: update jest config

matyax commented:
@itsmylife I took the liberty to update `jest.config.js` to properly set `transformIgnorePatterns`.

With that, @sd2k Jest is failing when Augurs uses `import.meta`. Do you think it would be possible to release a new version that does not rely on that? Otherwise, projects will have to adapt and install babel plugins to work around it, and that can be specially annoying in big repos like grafana/grafana. Thanks!

sd2k committed:
chore: create mock for @bsull/augurs (#92156)

chore: create mock for bsull/augurs

@bsull/augurs assumes it will be running as an ESM, not
a CommonJS module, so can't be loaded by Jest (specifically
because it contains a reference to import.meta.url).

This PR provides a mock implementation which gets tests passing
again.

Ideally we'd be able to load the actual @bsull/augurs module
in tests so this is just a stopgap really, until a better
solution appears.

sd2k commented:
It looks like the augurs failing tests are sorted now, but there are a couple of failing tests related to stddev / alphabetical sorting.

itsmylife committed:
Merge branch 'refs/heads/main' into ismail/em/implement-changepoint-sorting

itsmylife committed:
fix unit tests

sd2k commented on public/app/features/trails/services/sorting.ts:
```diff
@@ -0,0 +1,137 @@
+  const nTimestamps = joinedSeries[0].values.length;
+  const points = new Float64Array(joinedSeries.flatMap((series) => series.values as number[]));
+
+  try {
+    const detector = OutlierDetector.dbscan({ sensitivity: 0.4 }).preprocess(points, nTimestamps);
+    outliers = detector.detect();
+  } catch (e) {
+    console.error(e);
+  }
+};
+
+let outliers: OutlierOutput | undefined = undefined;
```
From what I can tell this is the same way that it's done in Explore Logs but I'm a bit confused about why we're using a global for `outliers` here rather than just a local var passed to `initOutlierDetector` / `calculateOutlierValue`?

It also looks like that could cause a stale value of `outliers` which could crash if:

1. The user selects outlier sorting and it works correctly (`outliers` will not be `undefined`)
2. They switch to another view
3. They select outlier sorting again and it fails, so that `outliers` is never overwritten but is still not `undefined`
4. `calculateOutlierValue` doesn't know that `outliers` is stale, and tries to index the old value of `outliers.seriesResults` with an invalid value of `index`

Perhaps I'm missing something though!

cc @matyax

matyax replied:
You make a great point here @sd2k . I would say it shouldn't be a problem because if it panics and then throws, it should be catched by the try/catch on line 25, which falls back to stddev.

itsmylife committed:
Merge branch 'refs/heads/main' into ismail/em/implement-changepoint-sorting

matyax commented on public/app/features/trails/services/sorting.ts:
```diff
@@ -0,0 +1,137 @@
+    const detector = OutlierDetector.dbscan({ sensitivity: 0.4 }).preprocess(points, nTimestamps);
+    outliers = detector.detect();
+  } catch (e) {
+    console.error(e);
```
Follwing @sd2k comment, I would initialize `ouliters = undefined` before the try/catch, or set to undefined if it throws. We should do the same in Explore Logs.

itsmylife replied:
Thanks for the suggestion @matyax 
here is the change https://github.com/grafana/grafana/pull/91189/commits/e072f811752731116f3a1fe2d490e9a52fe11ad9
I put it in `catch` block

itsmylife committed:
remove unused BreakdownScene.tsx

itsmylife committed:
set outliers as undefined if an error occurs

matyax reviewed (approved):
🚀

catherineymgui commented:
A small UI detail (separate from the larger outlier detection / widest spread discussion on the sorting discussion from Friday):

I noticed we are using inline labels. Could we standardize and use these vertical labels instead? I also noticed the label dropdown is missing 'By label'

Current:
<img width="1165" alt="Screenshot 2024-09-08 at 21 23 51" src="https://github.com/user-attachments/assets/9eba7bdd-1481-4841-95c9-8b6e643e60dc">

Proposed:

<img width="985" alt="Screenshot 2024-09-08 at 21 24 46" src="https://github.com/user-attachments/assets/53d61c1a-9f83-42a5-ba2c-ebc513db859c">

--- 

I think we can improve our copy too, but will wait for our discussion with the ML squad to think about this.

bohandley reviewed (approved):
LGTM! The sorting is working. Love the refactor in the breakdown and the select and search looks good.

bohandley commented:
There are no significants conflicts between this and the [OTel Explore metrics](https://github.com/grafana/grafana/pull/91221) work either! I am excited to see this merged and start using the refactored breakdown scene. 🚀

itsmylife committed:
Merge branch 'refs/heads/main' into ismail/em/implement-changepoint-sorting

itsmylife committed:
Add labels

itsmylife committed:
betterer

itsmylife committed:
reset event implemented

itsmylife committed:
fix controls positioning

itsmylife commented:
We decided to have a better implementation in terms of sorting and detecting the anomalies. So we close this PR in order to have a better implementation after we have a design doc for it in next quarter. Thanks for all the reviews and inputs.

catherineymgui commented:
Context: Based on further discussion, we decided to revisit this feature. So I am reopening this with the additional UX feedback:

1. I believe we discussed just using one type of ML sort (the outlier detection).

So we will have
- Detected outliers
- Alphabetical (A-Z)
- Alphabetical (Z-A)

2. The current description of detected outliers may be misleading. So I think we can update the copy and rename this 

Current:
- Detected outliers
- Order by the amount of detected outliers in the data
<img width="203" alt="Screenshot 2024-10-10 at 14 24 38" src="https://github.com/user-attachments/assets/b6fe8773-6bc4-4f16-a654-40f89d95f036">

Proposed:
- Outlier comparison (??)
- Compares and sorts values within the same label based on changes in behavior (such as spikes).
<img width="316" alt="Screenshot 2024-10-10 at 14 55 27" src="https://github.com/user-attachments/assets/97438e97-0107-4635-aeae-851f4d058e6f">

cc @sd2k @zhehao-grafana  if you had any other copy or naming suggestions!

3. Tooltip copy and width

Current:

<img width="192" alt="Screenshot 2024-10-10 at 14 24 30" src="https://github.com/user-attachments/assets/e26d3324-e0d8-4e20-8076-27dda94f1df9">

Proposed:
- Copy: Sorts values using standard or smart time series calculations.
- Width: Increase to 400px

zhehao-grafana commented:
> Outlier comparison (??)
What if we call it Outlier?

catherineymgui commented:
> > Outlier comparison (??)
> 
> What if we call it Outlier?

Ah yeah we can just call it 'Outliers'

sd2k commented:
Copy-wise, the proposed description sounds more like the [changepoint detection](https://grafana.com/developers/scenes/scenes-ml/changepoint-detection) (within series) algorithm rather than the [outlier detection](https://grafana.com/developers/scenes/scenes-ml/outlier-detection) (between series) algorithm. Closer to the truth would be something like:

> Prioritises (highlights?) values behaving differently to the others

There's surely a more precise way of putting it without it sounding too obtuse though 🤔

zhehao-grafana commented:
> Copy-wise, the proposed description sounds more like the [changepoint detection](https://grafana.com/developers/scenes/scenes-ml/changepoint-detection) (within series) algorithm rather than the [outlier detection](https://grafana.com/developers/scenes/scenes-ml/outlier-detection) (between series) algorithm. Closer to the truth would be something like:
> 
> > Prioritises (highlights?) values behaving differently to the others
> 
> There's surely a more precise way of putting it without it sounding too obtuse though 🤔

Can we say changepoint detection is more like outlier values and outlier detection is more like outlier series? If so, maybe the naming of Outlier Comparison should be Outlier Values?

sd2k commented:
> > Copy-wise, the proposed description sounds more like the [changepoint detection](https://grafana.com/developers/scenes/scenes-ml/changepoint-detection) (within series) algorithm rather than the [outlier detection](https://grafana.com/developers/scenes/scenes-ml/outlier-detection) (between series) algorithm. Closer to the truth would be something like:
> > > Prioritises (highlights?) values behaving differently to the others
> > 
> > 
> > There's surely a more precise way of putting it without it sounding too obtuse though 🤔
> 
> Can we say changepoint detection is more like outlier values and outlier detection is more like outlier series?

It's more or less correct to say that yeah. So far in our ML comms we've tried to stick with:

- 'outlier' being between series (i.e. one or more series behaving different to the rest), and
- 'anomalies' being within series (i.e. one or more time points behaving differently to the rest)

so have avoided using the term 'outlier' when we're talking about time points.

---

> If so, maybe the naming of Outlier Comparison should be Outlier Values?

Sounds reasonable to me. Maybe 'Outlying Values' depending on how we're phrasing things!

zhehao-grafana commented:
> Sounds reasonable to me. Maybe 'Outlying Values' depending on how we're phrasing things!

@catherineymgui let us know which one you prefer in terms of clarity!

catherineymgui commented:
Based on what @sd2k said, should we call this outlying series?

Name: Outlying Values **or** Outlying Series
Description: Prioritizes values **or** Prioritizes series that show distinct behavior from others within the same label

zhehao-grafana commented:
> Name: Outlying Values or Outlying Series
> Description: Prioritizes values or Prioritizes series that show distinct behavior from others within the same label

Let's go the following for now and unblock @itsmylife [we can always come back to tweak it]:
Name: Outlying Values
Description: Prioritizes values that show distinct behavior from others within the same label

itsmylife commented:
For the record, I am creating a new PR because that one has too many conflicts. Thanks for all the feedback. 
I'll link it with this one.

itsmylife committed:
Merge branch 'main' into ismail/em/implement-changepoint-sorting

# Conflicts:
#	.betterer.results
#	package.json
#	public/app/features/trails/Breakdown/ByFrameRepeater.tsx
#	public/app/features/trails/Breakdown/LabelBreakdownScene.tsx
#	public/app/features/trails/Breakdown/panelConfigs.ts
#	public/app/features/trails/interactions.ts
#	public/app/features/trails/services/levels.ts
#	public/app/features/trails/services/store.ts
#	public/app/features/trails/shared.ts
#	public/locales/en-US/grafana.json
#	public/locales/pseudo-LOCALE/grafana.json

itsmylife committed:
update augurs

itsmylife committed:
betterer

itsmylife committed:
i18n

itsmylife committed:
conflict fixes

itsmylife committed:
update texts

itsmylife commented:
@zhehao-grafana @catherineymgui I handled the requested changes. Please give this another go and then we will merge it.

zhehao-grafana commented:
Thanks Ismail for this! I played around this feature, and noticed one bug: it seems like there is a switch from the 1st graph to a different graph based on a selected label.


https://github.com/user-attachments/assets/0fcf34ac-ca47-45f5-a8fb-a2beabaf1a71

bohandley commented:
> Thanks Ismail for this! I played around this feature, and noticed one bug: it seems like there is a switch from the 1st graph to a different graph based on a selected label [It happened between 20s and 26s]
> 
>  Explore-Metrics-Sorting-Implementation.mp4

@zhehao-grafana @itsmylife I think this bug is related to updating the field config for the panel axis, so the graph looks different but it is the same, with a different axis. There is an event publish that is throttled for a second that when published, updates the axis.

The event
https://github.com/grafana/grafana/blob/ab7a96b6e0b01cca494e01ae8fadcd512a073e2c/public/app/features/trails/Breakdown/LabelBreakdownScene.tsx#L163-L168

The axis update
https://github.com/grafana/grafana/blob/ab7a96b6e0b01cca494e01ae8fadcd512a073e2c/public/app/features/trails/Breakdown/yAxisSyncBehavior.ts#L58-L83

This is definitely an issue but I don't think it is related to this PR. This code was written 3 months ago and merged in a previous PR. @itsmylife if you can confirm this, I think we can create a new issue for the bug @zhehao-grafana found and merge this 😃

zhehao-grafana commented:
> this bug is related to updating the field config for the panel axis, so the graph looks different but it is the same, with a different axis.

This makes sense now! Is this also related to this [bug](https://github.com/grafana/grafana/issues/91580)? Basically, the y axis scale is kept the same for all the label values

itsmylife committed:
Merge branch 'main' into ismail/em/implement-changepoint-sorting

# Conflicts:
#	public/app/features/trails/Breakdown/LabelBreakdownScene.tsx

sd2k commented:
Sorry for the late comment, just remembered to ask something: does this change mean the `@bsull/augurs` library is loaded as soon as the user visits any Grafana page, or is it delayed until they visit Explore Metrics? There's an initiative to reduce initial load times of Grafana and it's a fairly sizable WASM blob so it should probably be the latter.

(Separately, I'm going to work on reducing the size of the WASM soon, probably by splitting out the various pieces of functionality.)

itsmylife commented:
@sd2k It is initialized when a user enters the breakdown scene.

itsmylife commented:
@zhehao-grafana @bohandley This is not a bug actually. That is a feature. :D It's controlled by this https://github.com/grafana/grafana/blob/main/public/app/features/trails/Breakdown/yAxisSyncBehavior.ts

@darrenjaneczek has explained this to me. I was thinking it is a bug too. But if that's something we don't want we can remove that behaviour. But I agree with @bohandley, it is not related to this PR. 
So if those are clear I'd like to merge this PR.

itsmylife committed:
Merge branch 'main' into ismail/em/implement-changepoint-sorting

itsmylife committed:
linting

zhehao-grafana commented:
> @zhehao-grafana @bohandley This is not a bug actually. That is a feature. :D It's controlled by this https://github.com/grafana/grafana/blob/main/public/app/features/trails/Breakdown/yAxisSyncBehavior.ts
> 
> @darrenjaneczek has explained this to me. I was thinking it is a bug too. But if that's something we don't want we can remove that behaviour. But I agree with @bohandley, it is not related to this PR. So if those are clear I'd like to merge this PR.

Gotcha! We can address this as a separate issue from this PR, and feel free to merge it if everything else is ready 🚀

zhehao-grafana commented:
> @zhehao-grafana @bohandley This is not a bug actually. That is a feature. :D It's controlled by this https://github.com/grafana/grafana/blob/main/public/app/features/trails/Breakdown/yAxisSyncBehavior.ts
> 
> @darrenjaneczek has explained this to me. I was thinking it is a bug too. But if that's something we don't want we can remove that behaviour. But I agree with @bohandley, it is not related to this PR. So if those are clear I'd like to merge this PR.

Gotcha! We can address this as a separate issue from this PR, and feel free to merge it if everything else is ready 🚀

itsmylife committed:
one line of import
